File: system-test/java/src/main/java/com/optivem/shop/dsl/core/usecase/external/clock/ClockDsl.java
package com.optivem.shop.dsl.core.usecase.external.clock;

import com.optivem.shop.dsl.driver.port.external.clock.ClockDriver;
import com.optivem.shop.dsl.core.usecase.external.clock.usecases.GetTime;
import com.optivem.shop.dsl.core.usecase.external.clock.usecases.GoToClock;
import com.optivem.shop.dsl.core.usecase.external.clock.usecases.ReturnsTime;
import com.optivem.shop.dsl.common.Closer;
import com.optivem.shop.dsl.core.shared.UseCaseContext;

public class ClockDsl implements AutoCloseable {
    protected final ClockDriver driver;
    protected final UseCaseContext context;

    public ClockDsl(ClockDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToClock goToClock() {
        return new GoToClock(driver, context);
    }

    public ReturnsTime returnsTime() {
        return new ReturnsTime(driver, context);
    }

    public GetTime getTime() {
        return new GetTime(driver, context);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/clock/ClockDsl.ts
import type { ClockDriver } from '../../../../../driver/port/external/clock/clock-driver.js';
import type { UseCaseContext } from '../../../use-case-context.js';
import { GoToClock } from './usecases/GoToClock.js';
import { GetTime } from './usecases/GetTime.js';
import { ReturnsTime } from './usecases/ReturnsTime.js';

export class ClockDsl {
  constructor(
    private readonly driver: ClockDriver,
    private readonly context: UseCaseContext,
  ) {}

  goToClock(): GoToClock {
    return new GoToClock(this.driver, this.context);
  }

  getTime(): GetTime {
    return new GetTime(this.driver, this.context);
  }

  returnsTime(): ReturnsTime {
    return new ReturnsTime(this.driver, this.context);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}
